File: Medium/1717. Maximum Score From Removing Substrings.java
class Solution {
    public int maximumGain(String s, int x, int y) {
        final int lesser = Math.min(x, y);
        int aCnt = 0, bCnt = 0;
        int res = 0;
        for (final var c : s.toCharArray()) {
            if (c == 'a') {
                if (y > x && bCnt > 0) {
                    res += y;
                    bCnt--;
                } else {
                    aCnt++;
                }
            } else if (c == 'b') {
                if (x > y && aCnt > 0) {
                    res += x;
                    aCnt--;
                } else {
                    bCnt++;
                }
            } else {
                res += lesser * Math.min(aCnt, bCnt);
                aCnt = 0;
                bCnt = 0;
            }
        }
        res += lesser * Math.min(aCnt, bCnt);
        return res;
    }
}

File: Medium/1717. Maximum Score From Removing Substrings.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        """
        Calculate the maximum score by removing substrings "ab" and "ba" from the input string.

        :param s: Input string
        :param x: Score for removing "ab"
        :param y: Score for removing "ba"
        :return: Maximum score achievable
        """
        a = b = max_points = 0
        for c in s:
            if c == 'b':
                if x >= y and a:
                    a -= 1
                    max_points += x
                else:
                    b += 1
            elif c == 'a':
                if y > x and b:
                    b -= 1
                    max_points += y
                else:
                    a += 1
            else:
                max_points += min(a, b) * (y if x >= y else x)
                a = b = 0
        return max_points + min(a, b) * (y if x >= y else x)